File: demo/pythonCodes/mysqrt.py
# this is a demo code for implementing the Babylonian Algorithm
# to compute the square root of a number.
#
# we have demonstrated the developping process in the class
#
def mysqrt(S,x0=1.,tol=1e-5,debug=False):
    #Input:
    #       S: the number we want to compute its square root
    #      x0: initial guess. We have provided a default value, so it's optional
    #      tol: tolerance. Default values is provided as tol=1e-5
    #debug: a True/False flag to indicate if to turn on the debugging mode
    
    # check if input is valid
    if S<0.:
        print("Error: we can only compute square root of a positive number\n")
        return

    #max number of iteration allowed
    Nmax=100;
    xold=x0;
    for n in range(Nmax):
        xnew = 0.5*(xold+S/xold)
        if debug:
            print("xnew=",xnew,"at step=",n)
            #check stopping criteria
        if abs(xnew-xold)<tol:
            return xnew
        #update xnew for next iteration
        xold=xnew
        
    print("Error: The algorithm failed to converge to tol after the max number of iterations reached\n")

    return

File: demo/matlabCodes/mysqrt.m
function r=mysqrt(S,x0,tol,debug)
% Input:
%   S: value we want to compute its square root
%   x0: initial guess
%   tol: tolerance 
%   debug: a flag to turn on/off the debug mode
% Output:
%   r: result, i,e., sqrt(S)

% check if input is valid
if S<0
    fprintf("Error: we can only compute square root of a positive number\n");
    return
end

%max number of iteration allowed
Nmax=100;
xold=x0;
for n = 1:Nmax
    xnew = 0.5*(xold+S/xold);
    if debug
        fprintf(sprintf("xnew=%f at step= %d \n",xnew,n));
    end
    %check stopping criteria
    if abs(xnew-xold)<tol
         r=xnew;
         return
    end
    %update xnew for next iteration
    xold=xnew;
end

fprintf("Error: The algorithm failed to converge to tol after the max number of iterations reached\n")


end